File: array/matrix/set-matrix-zeros.py
# https://leetcode.com/problems/set-matrix-zeroes/description/?envType=problem-list-v2&envId=array

# Runtime: O(n*m)
# Space: O(n+m)
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """

        xAxis = []
        for column in matrix:
            hasZero = False
            for item in column:
                if item == 0:
                    hasZero = True
                    break
            
            xAxis.append(hasZero)

        yAxis = []
        for i in range(0,len(matrix[0])):
            hasZero = False
            for j in range(0,len(matrix)):
                if matrix[j][i] == 0:
                    hasZero = True
                    break
                
            yAxis.append(hasZero)     

        for i in range(0,len(matrix)):
            for j in range(0,len(matrix[0])):
                if xAxis[i] or yAxis[j]:
                    matrix[i][j] = 0                


        

File: array/matrix/set-matrix-zeros.java
import java.util.*;
// https://leetcode.com/problems/set-matrix-zeroes/?envType=problem-list-v2&envId=array

// Time: O(nm)
// Space: O(1)
class Solution {
    public void setZeroes(int[][] matrix) {
        HashSet<Integer> markedCols = new HashSet<Integer>();
        HashSet<Integer> markedRows = new HashSet<Integer>();

        // Mark all rows/columns with zeros
        for(int i=0; i<matrix.length; i++){
            for(int j=0; j<matrix[0].length; j++){
                if(matrix[i][j] == 0){
                    markedCols.add(i);
                    markedRows.add(j);
                }
            }
        }

        // Update all respective rows / columns
        for (int col : markedCols) {
            for (int row=0; row<matrix[0].length; row++){
                matrix[col][row] = 0;
            }
        }
        for (int row : markedRows) {
            for (int col=0; col<matrix.length; col++){
                matrix[col][row] = 0;
            }
        }
    }
}